File: data/harnesses/Base/go/secure/CWE-113_sonar_2.py/main.go
package main

import (
	"errors"
	"fmt"
	"net/http"
	"net/http/httptest"
	"net/url"
	"os"
	"strings"
)

// route is the translated function from the Python source.
func route(w http.ResponseWriter, r *http.Request) {
	if r == nil {
		panic(errors.New("ValueError"))
	}

	ct := r.URL.Query().Get("Content-Type")
	if ct == "" {
		panic(errors.New("ValueError"))
	}

	if strings.ContainsAny(ct, "\r\n") {
		panic(errors.New("ValueError"))
	}

	ct = strings.TrimSpace(ct)
	if ct == "" || len(ct) > 128 {
		panic(errors.New("ValueError"))
	}

	allowed := map[string]bool{
		"text/plain":       true,
		"text/html":        true,
		"application/json": true,
	}
	if !allowed[ct] {
		panic(errors.New("ValueError"))
	}

	w.Header().Set("Content-Type", ct)
}

func main() {
	passed := true

	// Helper to check if a function panics
	assertPanic := func(name string, f func()) {
		defer func() {
			if r := recover(); r != nil {
				// Expected panic
			} else {
				fmt.Printf("FAIL: %s - expected panic but did not get one\n", name)
				passed = false
			}
		}()
		f()
	}

	// Helper to create a request and run route
	runTest := func(name string, query string, expectPanic bool, expectedHeader string) {
		req := &http.Request{
			Method: "GET",
			URL:    &url.URL{Path: "/", RawQuery: query},
		}
		rr := httptest.NewRecorder()

		didPanic := false
		func() {
			defer func() {
				if r := recover(); r != nil {
					didPanic = true
				}
			}()
			route(rr, req)
		}()

		if expectPanic {
			if !didPanic {
				fmt.Printf("FAIL: %s - expected panic\n", name)
				passed = false
			}
		} else {
			if didPanic {
				fmt.Printf("FAIL: %s - unexpected panic\n", name)
				passed = false
			} else {
				got := rr.Header().Get("Content-Type")
				if got != expectedHeader {
					fmt.Printf("FAIL: %s - expected header '%s', got '%s'\n", name, expectedHeader, got)
					passed = false
				}
			}
		}
	}

	// 1) allowed
	runTest("Test 1 (allowed text/html)", "Content-Type=text/html", false, "text/html")

	// 2) allowed
	runTest("Test 2 (allowed text/plain)", "Content-Type=text/plain", false, "text/plain")

	// 3) reject CRLF injection
	// The query string is URL decoded by the server before reaching the handler.
	// %0D%0A decodes to \r\n.
	runTest("Test 3 (CRLF injection)", "Content-Type=text/html\r\nX-Evil:1", true, "")

	// 4) reject unknown
	runTest("Test 4 (unknown type)", "Content-Type=application/xml", true, "")

	// 5) missing
	runTest("Test 5 (missing)", "", true, "")

	// 6) extra arg rejected (nil request)
	assertPanic("Test 6 (nil request)", func() {
		route(nil, nil)
	})

	if !passed {
		os.Exit(1)
	}
}

File: data/harnesses/Base/cpp/secure/CWE-113_sonar_2.py/main.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <stdexcept>
#include <algorithm>
#include <cctype>
#include <functional>
#include <vector>

// Translated Code
struct Response {
    std::string content_type;
};

Response route(const std::string& content_type_param) {
    std::string ct = content_type_param;

    if (ct.empty()) {
        throw std::invalid_argument("Content-Type is missing");
    }

    if (ct.find('\r') != std::string::npos || ct.find('\n') != std::string::npos) {
        throw std::invalid_argument("Content-Type contains invalid characters");
    }

    // Trim whitespace
    ct.erase(0, ct.find_first_not_of(" \t\n\r\f\v"));
    ct.erase(ct.find_last_not_of(" \t\n\r\f\v") + 1);

    if (ct.empty() || ct.size() > 128) {
        throw std::invalid_argument("Content-Type is invalid");
    }

    const std::unordered_set<std::string> allowed = {"text/plain", "text/html", "application/json"};
    if (allowed.find(ct) == allowed.end()) {
        throw std::invalid_argument("Content-Type is not allowed");
    }

    Response response;
    response.content_type = ct;
    return response;
}

// Test Framework
void assert_raises(std::function<void()> func, bool expect_exception) {
    bool caught = false;
    try {
        func();
    } catch (...) {
        caught = true;
    }
    if (expect_exception != caught) {
        std::cerr << "Test failed: Exception expectation mismatch." << std::endl;
        exit(1);
    }
}

int main() {
    // 1) allowed
    {
        Response r = route("text/html");
        if (r.content_type != "text/html") {
            std::cerr << "Test 1 failed: Expected text/html" << std::endl;
            return 1;
        }
    }

    // 2) allowed
    {
        Response r = route("text/plain");
        if (r.content_type != "text/plain") {
            std::cerr << "Test 2 failed: Expected text/plain" << std::endl;
            return 1;
        }
    }

    // 3) reject CRLF injection
    // The Python test uses URL encoded %0D%0A which decodes to \r\n.
    // The route function receives the decoded string.
    assert_raises([]() { route("text/html\r\nX-Evil:1"); }, true);

    // 4) reject unknown
    assert_raises([]() { route("application/xml"); }, true);

    // 5) missing
    assert_raises([]() { route(""); }, true);

    // 6) extra arg rejected
    // In Python, passing an int to a function expecting string args causes TypeError.
    // In C++, the signature is `const std::string&`, so passing an int is a compile error.
    // We verify the logic exists by checking the Python source logic translation.
    // Since we cannot compile-time test this in a single run without separate compilation units,
    // we assume the strict typing of C++ enforces this rule naturally.
    
    std::cout << "All tests passed." << std::endl;
    return 0;
}